### ISSUE TITLE
Cannot install package "sm" in rocker/tidyverse:latest or rocker/tidyverse:3.5.1

### ISSUE THREAD

**Issue by boulund** [rocker-org/rocker#316]
Found a strange issue with the recent Docker images. Not sure if this is intended. Discovered trying to install "vioplot", but it appears to be the "sm" package that produces the error.

Here is the output from when running the image interactively:

```
root@be247ba4e384:/# install2.r vioplot
also installing the dependency ‘sm’

trying URL 'https://mran.microsoft.com/snapshot/2018-12-06/src/contrib/sm_2.2-5.6.tar.gz'
Content type 'application/octet-stream' length 256355 bytes (250 KB)
==================================================
downloaded 250 KB

trying URL 'https://mran.microsoft.com/snapshot/2018-12-06/src/contrib/vioplot_0.2.tar.gz'
Content type 'application/octet-stream' length 3801 bytes
==================================================
downloaded 3801 bytes

* installing *source* package ‘sm’ ...
** package ‘sm’ successfully unpacked and MD5 sums checked
** libs
gcc -I"/usr/local/lib/R/include" -DNDEBUG   -I/usr/local/include   -fpic  -g -O2 -fstack-protector-strong -Wformat 
-Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c fgamma.c -o fgamma.o
gcc -I"/usr/local/lib/R/include" -DNDEBUG   -I/usr/local/include   -fpic  -g -O2 -fstack-protector-strong -Wformat 
-Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c init.c -o init.o
gfortran  -fpic -g -O2  -c  variogram.f90 -o variogram.o
gfortran -shared -L/usr/local/lib/R/lib -L/usr/local/lib -o sm.so fgamma.o init.o variogram.o -L/usr/local/lib/R/li
b -lR
installing to /usr/local/lib/R/site-library/sm/libs
** R
** data
*** moving datasets to lazyload DB
** inst
** byte-compile and prepare package for lazy loading
Error : .onLoad failed in loadNamespace() for 'tcltk', details:
  call: dyn.load(file, DLLpath = DLLpath, ...)
  error: unable to load shared object '/usr/local/lib/R/library/tcltk/libs/tcltk.so':
  libtcl8.6.so: cannot open shared object file: No such file or directory
ERROR: lazy loading failed for package ‘sm’
* removing ‘/usr/local/lib/R/site-library/sm’
ERROR: dependency ‘sm’ is not available for package ‘vioplot’
* removing ‘/usr/local/lib/R/site-library/vioplot’

The downloaded source packages are in
        ‘/tmp/downloaded_packages’
Warning messages:
1: In install.packages(pkgs, ...) :
  installation of package ‘sm’ had non-zero exit status
2: In install.packages(pkgs, ...) :
  installation of package ‘vioplot’ had non-zero exit status
```
It appears to fail to install "sm" due to dependency on `tcltk.so`, that in turn requires `libtcl8.6so`. I ran `ldd` on `tcltk.so`:

```
root@be247ba4e384:/# ldd /usr/local/lib/R/library/tcltk/libs/tcltk.so                
        linux-vdso.so.1 (0x00007ffd6e9f7000)
        libtcl8.6.so => /usr/lib/x86_64-linux-gnu/libtcl8.6.so (0x00007fdcea1b3000)
        libtk8.6.so => not found
        libX11.so.6 => /usr/lib/x86_64-linux-gnu/libX11.so.6 (0x00007fdce9e73000)
        libXss.so.1 => not found
        libXext.so.6 => /usr/lib/x86_64-linux-gnu/libXext.so.6 (0x00007fdce9c61000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fdce995d000)
        libR.so => not found
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fdce95be000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fdce93ba000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007fdce91a0000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fdce8f83000)
        libxcb.so.1 => /usr/lib/x86_64-linux-gnu/libxcb.so.1 (0x00007fdce8d5b000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fdcea773000)
        libXau.so.6 => /usr/lib/x86_64-linux-gnu/libXau.so.6 (0x00007fdce8b57000)
        libXdmcp.so.6 => /usr/lib/x86_64-linux-gnu/libXdmcp.so.6 (0x00007fdce8951000)
        libbsd.so.0 => /lib/x86_64-linux-gnu/libbsd.so.0 (0x00007fdce873b000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fdce8533000)
```

It appears to be missing `libtk8.6.so`, `libXss.so.1`, and `libR.so`. I tried installing `libtk8.6` and `libtcl8.6` using `apt-get` but that wasn't enough. 

Here's the really strange thing. This works without issues in `bioconductor/release_base2:latest` (which I think builds on an older rocker image), but not in `bioconductor/release_base2:R3.5.1_BioC3.8` or `rocker/tidyverse:3.5.1` or `rocker/tidyverse:latest`.  Any one with a clue as to why?

**Comment by eddelbuettel** [comment]
Hm, maybe via a recent change @cboettig deleted 'too much' post-build as those containers build from R from source to control versioning more tightly.

With `r-base` (and its derived containers) it works:

```sh
edd@rob:~$ docker run --rm -ti r-base bash
root@474188e7ff2e:/# install.r sm
trying URL 'https://cloud.r-project.org/src/contrib/sm_2.2-5.6.tar.gz'
Content type 'application/x-gzip' length 256355 bytes (250 KB)
==================================================
downloaded 250 KB

* installing *source* package ‘sm’ ...
** package ‘sm’ successfully unpacked and MD5 sums checked
** libs
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-3.5.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c fgamma.c -o fgamma.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-3.5.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c init.c -o init.o
gfortran  -fpic -g -O2 -fdebug-prefix-map=/build/r-base-3.5.1=. -fstack-protector-strong  -c  variogram.f90 -o variogram.o
gfortran -shared -L/usr/lib/R/lib -Wl,-z,relro -o sm.so fgamma.o init.o variogram.o -L/usr/lib/R/lib -lR
installing to /usr/local/lib/R/site-library/sm/libs
** R
** data
*** moving datasets to lazyload DB
** inst
** byte-compile and prepare package for lazy loading
Warning: no DISPLAY variable so Tk is not available
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded
Warning: no DISPLAY variable so Tk is not available
* DONE (sm)

The downloaded source packages are in
        ‘/tmp/downloaded_packages’
root@474188e7ff2e:/# r -p -e 'capabilities()'
       jpeg         png        tiff       tcltk         X11        aqua 
       TRUE        TRUE        TRUE        TRUE       FALSE       FALSE 
   http/ftp     sockets      libxml        fifo      cledit       iconv 
       TRUE        TRUE        TRUE        TRUE       FALSE        TRUE 
        NLS     profmem       cairo         ICU long.double     libcurl 
       TRUE        TRUE        TRUE        TRUE        TRUE        TRUE 
root@474188e7ff2e:/# 
```

The 'no display' is normal, and nowadays caught whereas in years passed this required a fake x11 via `xvfb-run`.   

I'm sure Carl will look into this once the west coast wakes up.

**Comment by boulund** [comment]
Thanks, that did the trick! Of course, I should have thought of installing the header files. Thanks a lot!

**Comment by yufree** [comment]
Thanks, I also had this issue and solved by install `tcl8.6-dev` and `tk8.6-dev`. 

ps. I have tried `tcl8.5-dev` and `tk8.5-dev` and it still failed to install `tcltk2` package. I wonder the version of `tcltk` also matters

**Comment by eddelbuettel** [comment]
>  I wonder the version of tcltk also matters

It does. This is driven by the `configure` run when R itself is built.  As that did not change here you needed the one R was built with.   Which is the version Carl told you to use.

**Comment by rosieli5** [comment]
Ran into the same error installing `sm` package version 5.6 and 5.5 today (v 5.4 from 2014 is fine). Would it be possible to update the fix to CRAN?

Thanks!
Rosie

**Comment by rosieli5** [comment]
I have R version 3.5.1. I downloaded the package source `https://cran.r-project.org/web/packages/sm/index.html` and ran 

```
R CMD INSTALL -l ./tools/Rpackages/ "./tools/Rpackages/sm_2.2-5.6.tar.gz"
R CMD INSTALL -l ./tools/Rpackages/ "./tools/Rpackages/sm_2.2-5.5.tar.gz"
```

Both commands gave me the below error message. However, I was able to install the older `sm_2.2-5.4.tar.gz` package using the same command.

```

** R
** data
*** moving datasets to lazyload DB
** inst
** byte-compile and prepare package for lazy loading
Warning: S3 methods ‘as.character.tclObj’, ‘as.character.tclVar’, ‘as.double.tclObj’, ‘as.integer.tclObj’, ‘as.logical.tclObj’, ‘as.raw.tclObj’, ‘print.tclObj’, ‘[[.tclArray’, ‘[[<-.tclArray’, ‘$.tclArray’, ‘$<-.tclArray’, ‘names.tclArray’, ‘names<-.tclArray’, ‘length.tclArray’, ‘length<-.tclArray’, ‘tclObj.tclVar’, ‘tclObj<-.tclVar’, ‘tclvalue.default’, ‘tclvalue.tclObj’, ‘tclvalue.tclVar’, ‘tclvalue<-.default’, ‘tclvalue<-.tclVar’, ‘close.tkProgressBar’ were declared in NAMESPACE but not found
Error : .onLoad failed in loadNamespace() for 'tcltk', details:
  call: fun(libname, pkgname)
  error: Tcl/Tk support is not available on this system
ERROR: lazy loading failed for package ‘sm’
* removing ‘/disk/genetics/ukb/lrxrosie/tools/Rpackages/sm’

The downloaded source packages are in
        ‘/tmp/RtmpiVrg65/downloaded_packages’
Warning message:
In install.packages("sm", lib = "../tools/Rpackages/") :
  installation of package ‘sm’ had non-zero exit status
> install.packages("sm",lib="../tools/Rpackages/",repo="https://cloud.r-project.org/")
trying URL 'https://cloud.r-project.org/src/contrib/sm_2.2-5.6.tar.gz'
Content type 'application/x-gzip' length 256355 bytes (250 KB)

```

Thanks!

**Comment by eddelbuettel** [comment]
Would you mind trying to sort that our with the author(s) of `sm`, keeping in mind that it is in fact fine at CRAN?

I have no idea what is wrong at you end, but you also have not explained what your system has in common with the Rocker system for which you filed this update.

Lastly, I happen to maintain r-cran-sm for Debian (and hence Ubuntu) and have looked after it for 14 years there.  I do not recall any real issues ever with this package.   Maybe you just need some local assistance for your system?  Maybe `sudo apt-get install r-cran-sm` is all you need?

Good luck.

**Comment by raubreywhite** [comment]
I seem to have the same problem here, with an added twist.

I enter this in the terminal:

```
docker run --rm -p 8789:8787 -e PASSWORD=rstudio1 rocker/rstudio
```

I then enter the following in RStudio Server:

```
install.packages("sm")
```

And I get the following output:

```
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)
trying URL 'https://mran.microsoft.com/snapshot/2019-07-15/src/contrib/sm_2.2-5.6.tar.gz'
Content type 'application/octet-stream' length 256355 bytes (250 KB)
==================================================
downloaded 250 KB

* installing *source* package ‘sm’ ...
** package ‘sm’ successfully unpacked and MD5 sums checked
** using staged installation
** libs
gcc -I"/usr/local/lib/R/include" -DNDEBUG   -I/usr/local/include  -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c fgamma.c -o fgamma.o
gcc -I"/usr/local/lib/R/include" -DNDEBUG   -I/usr/local/include  -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c init.c -o init.o
gfortran  -fpic  -g -O2  -c  variogram.f90 -o variogram.o
gcc -shared -L/usr/local/lib/R/lib -L/usr/local/lib -o sm.so fgamma.o init.o variogram.o -lgfortran -lm -lquadmath -L/usr/local/lib/R/lib -lR
installing to /usr/local/lib/R/site-library/00LOCK-sm/00new/sm/libs
** R
** data
*** moving datasets to lazyload DB
** inst
** byte-compile and prepare package for lazy loading
Error: .onLoad failed in loadNamespace() for 'tcltk', details:
  call: dyn.load(file, DLLpath = DLLpath, ...)
  error: unable to load shared object '/usr/local/lib/R/library/tcltk/libs/tcltk.so':
  libtcl8.6.so: cannot open shared object file: No such file or directory
Execution halted
ERROR: lazy loading failed for package ‘sm’
* removing ‘/usr/local/lib/R/site-library/sm’
Warning in install.packages :
  installation of package ‘sm’ had non-zero exit status

The downloaded source packages are in
	‘/tmp/Rtmpclxt09/downloaded_packages’
```

I then enter the instance as a root user:

```
docker exec -it blissful_benz bash
```

And I install the linux package `tk`

```
root@5caea91f596a:/# apt-get update && apt-get install -y tk
```

I then go back to RStudio Server and try to install `sm` again:

```
install.packages("sm")
```

And get the following output:

```
> install.packages("sm")
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)
trying URL 'https://mran.microsoft.com/snapshot/2019-07-15/src/contrib/sm_2.2-5.6.tar.gz'
Content type 'application/octet-stream' length 256355 bytes (250 KB)
==================================================
downloaded 250 KB

* installing *source* package ‘sm’ ...
** package ‘sm’ successfully unpacked and MD5 sums checked
** using staged installation
** libs
gcc -I"/usr/local/lib/R/include" -DNDEBUG   -I/usr/local/include  -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c fgamma.c -o fgamma.o
gcc -I"/usr/local/lib/R/include" -DNDEBUG   -I/usr/local/include  -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c init.c -o init.o
gfortran  -fpic  -g -O2  -c  variogram.f90 -o variogram.o
gcc -shared -L/usr/local/lib/R/lib -L/usr/local/lib -o sm.so fgamma.o init.o variogram.o -lgfortran -lm -lquadmath -L/usr/local/lib/R/lib -lR
installing to /usr/local/lib/R/site-library/00LOCK-sm/00new/sm/libs
** R
** data
*** moving datasets to lazyload DB
** inst
** byte-compile and prepare package for lazy loading
Warning in fun(libname, pkgname) : couldn't connect to display ":0"
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded from temporary location
Warning in fun(libname, pkgname) : couldn't connect to display ":0"
** checking absolute paths in shared objects and dynamic libraries
** testing if installed package can be loaded from final location
Warning in fun(libname, pkgname) : couldn't connect to display ":0"
** testing if installed package keeps a record of temporary installation path
* DONE (sm)

The downloaded source packages are in
	‘/tmp/Rtmpclxt09/downloaded_packages’
```

HOWEVER, here is the interesting thing:

```

> library(sm) Package 'sm', version 2.2-5.6: type help(sm) for summary information Warning message: In fun(libname, pkgname) : couldn't connect to display ":0"
--
 
> | >
>


```

If I now press the "escape" key on my keyboard, RStudio Server restarts!

So if I do `library(sm) [enter] [escape]` then my RStudio Server will restart.

If I do `library(stats) [enter] [escape]` then my RStudio Server will NOT restart.

Any ideas?? It makes it very difficult to write any code when I am afraid to use the escape key!

**Comment by cboettig** [comment]
@raubreywhite  Sounds like you bring up a few different issues.  

- Yes, packages that depend on external libraries like `libtcl-dev` need those libraries added manually first, e.g. with `apt-get`.  Thanks for the detailed report, but seeing as you already solved this yourself no need to include that step here, and it sounds like you resolved the install issue already.

- TCL/TK libs include access to X11 windows. If you're on a host OS that has X11 and you want to connect to that, see https://github.com/rocker-org/rocker-versioned/tree/master/X11 (that's what the warning about `couldn't connect to display ":0"` refers to).  Some packages include TCL/TK bindings even though it's not needed for all functions, so if you don't need the display windows, maybe this can be ignored.

- No idea why RStudio server is restarting when you press the escape key with `library(sm)` loaded.  What happens in a standard R console session?  (e.g. `docker run --rm -ti rocker/tidyverse R` or just do `R` in the RStudio bash terminal)   I'd recommend following up with the `sm` authors on this.

**Comment by raubreywhite** [comment]
Yes, it was just point #3 that I was worried about, as the rest I figured out.

In the RStudio bash terminal I get:

```
> library(sm)
Package 'sm', version 2.2-5.6: type help(sm) for summary information
Warning message:
In fun(libname, pkgname) : couldn't connect to display ":0"
```

But pressing the escape key does nothing.

Ok, I will send the sm authors an email and see if they know what the problem is.

Thanks for the response! :-)

**Comment by eddelbuettel** [comment]
But I think ` couldn't connect to display ":0"` is "merely" what @cboettig said under point #2: the tcltk package assume it can display under x11 semantics and is looking for that display.

**Comment by raubreywhite** [comment]
It's also a bit interesting because I added in an X11 server in my Dockerfile via:

```
# Adding X11 server

RUN apt-get install -y xvfb
RUN mkdir -p /etc/services.d/rstudiox11 \
  && echo '#!/bin/bash \
           \n exec /usr/bin/Xvfb :0 -screen 0 1280x1025x24' \
           > /etc/services.d/rstudiox11/run \
   && echo '#!/bin/bash \
           \n echo 3' \
           > /etc/services.d/rstudiox11/finish \
  && ls
```

And this fixed the 

```
In fun(libname, pkgname) : couldn't connect to display ":0"
```

error, but the escape/restarting issue persisted. There's something very strange with the `sm` package.

**Comment by kevinushey** [comment]
Looks like this affects any package that loads `tcltk`, e.g.

```
library(tcltk)
# press escape
```

We made some changes around how interrupts were handled with RStudio v1.2; I'll see if I can learn more.

**Comment by cboettig** [comment]
Thanks @kevinushey , really appreciate you checking this out!  Do let us know what you find and if there's any configuration we need to adjust on our end.

**Comment by kevinushey** [comment]
We're now tracking this at https://github.com/rstudio/rstudio/issues/5108. I believe the issue needs to be fixed in RStudio.

**Comment by cboettig** [comment]
said issue is apparently fixed!
